### ISSUE TITLE
speed up f16 on CUDA with hmul2, hadd2, etc.

### ISSUE THREAD

**Issue by gaika** [arrayfire/arrayfire#2752]
f16 CUDA can be faster with specialized half ops that operate on vector inputs. Correct me if I'm wrong, but I don't see them being used in the source code.

**Comment by umar456** [comment]
Yes, this is correct. You could get 2x speedup over regular ops with this change. Unfortunately, this will require a good bit of work to get it work correctly and there are very few functions that would take advantage of these ops because most kernels are memory bound.

**Comment by WilliamTambellini** [comment]
Note: kind of related to 
https://github.com/arrayfire/arrayfire/issues/2635
@umar456  would using hmul2/hadd2 jit kernels  speed up f16 gemv and/or non multiple of 4 gemm ?

**Comment by umar456** [comment]
Its unlikely that we will speedup JIT kernels. I suspect most of those kernels are memory bound. gemm is already optimized by cublas and other libraries so those times will not improve. I have tested this with reduce and there wasn't a huge difference between the two approaches.

**Comment by WilliamTambellini** [comment]
Hum, I would prefer to keep that item opened : gemv perf is important for me. 
Note : related change :  
https://gitlab.com/libeigen/eigen/-/merge_requests/51
